### PR TITLE
Add ability to save inner pixel triplets as well (if satisfy a condition)

### PR DESCRIPTION
**Pull request by VinInn**
I have added the ability to save inner pixel triplets as found by the CA if they satisfy some condition
(in this case if they point to the "hole" between barrel and endcap.
This is simply controlled by
```process.pixelTracksHitQuadruplets.minHitsPerNtuplet = 3```
(4 is default)

At the same time I implemented independent fitting of each tuple multiplicity including pentuplets.
Pentuplets can still be fitted as "inner quadruplets" setting
```process.pixelTracksHitQuadruplets.fit5as4 = True`` that is default.

therefore default is previous behavior

This PR is on top of #269 

here is some Physics performance result (MTV)
Single muon 
http://innocent.home.cern.ch/innocent/RelVal/SingleMu50_gpuBL345/plots_pixel_pixel/effandfakePtEtaPhi.pdf
ttbar PU50
http://innocent.home.cern.ch/innocent/RelVal/pixOnlyPU50_gpuBLTri/plots_pixel_pixel/effandfakePtEtaPhi.pdf
http://innocent.home.cern.ch/innocent/RelVal/pixOnlyPU50_gpuBLTri/plots_pixelFromPV_pixelFromPV/effandfakePtEtaPhi.pdf (only track associated to PV, fake include PU)
clearly the fake rate (computed w/r/t the signal) w/o vertex constraint is unbearable while looks manageable with constraint.
The interest of having it w/o vertex constraint relies of course in having a better charge track PU estimation in that area, in such a case of course the fake rate should be computed w/r/t the total number of tracks, not just signal.

One case use this setup to study pentuplet fitting: enough to set 
```process.pixelTracksHitQuadruplets.minHitsPerNtuplet = 5```
in http://innocent.home.cern.ch/innocent/RelVal/SingleMu50_gpu5only/plots_pixel_pixel/effandfakePtEtaPhi.pdf
http://innocent.home.cern.ch/innocent/RelVal/SingleMu50_gpu5only/plots_pixel_pixel/resolutionsEta.pdf
One can observe the result applied to single muon for the 4 cases: Riemann/BrokenLine fitting 5 or 4 hits.
The statistics is not enormous still the results are a bit deceiving: the resolution seems worse and Riemann 5 as 5 has inefficiencies due to failed fit (in fact using debug dump one can observe many failed fit (very large chi2, NaNs).
Failed fits seem to affect also BrokenLine in ttbar PU50.
More studies are required (and this PR is a starting point)

### CODE REVIEW

File: RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h [hunk 1]
```diff
@@ -159,19 +175,23 @@ class GPUCACell {
...
       }
     } else {  // if long enough save...
       if ((unsigned int)(tmpNtuplet.size()) >= minHitsPerNtuplet-1) {
-        hindex_type hits[6]; auto nh=0U;
-        for (auto c : tmpNtuplet) hits[nh++] = cells[c].theInnerHitId;
-        hits[nh] = theOuterHitId; 
-        uint16_t it = foundNtuplets.bulkFill(apc,hits,tmpNtuplet.size()+1);
-        for (auto c : tmpNtuplet) cells[c].theTracks.push_back(it);
+        // triplets accepted only pointing to the hole
+        if (tmpNtuplet.size()>=3 || hole(hh, cells[tmpNtuplet[0]])) {
```
Comment (VinInn) [inline RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h]: This additional runtime condition (that for minHitsPerNtuplet==4 is always true and should not require hole to be invoked) seems NOT to be particularly cheap:
   * before
```12.44%  2.99829s     20000  149.91us  26.912us  617.05us  kernel_find_ntuplets(GPUCACell*, unsigned int const *, HistoContainer<unsigned int, unsigned int=10000, unsigned int=50000, unsigned int=32, unsigned short, unsigned int=1>*, AtomicPairCounter*, unsigned int)```
   * after
```12.44%  2.99829s     20000  149.91us  26.912us  617.05us  kernel_find_ntuplets(GPUCACell*, unsigned int const *, HistoContainer<unsigned int, unsigned int=10000, unsigned int=50000, unsigned int=32, unsigned short, unsigned int=1>*, AtomicPairCounter*, unsigned int)```

File: RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h [hunk 2]
```diff
@@ -159,19 +175,23 @@ class GPUCACell {
...
+          hindex_type hits[6]; auto nh=0U;
+          for (auto c : tmpNtuplet) hits[nh++] = cells[c].theInnerHitId;
+          hits[nh] = theOuterHitId; 
+          uint16_t it = foundNtuplets.bulkFill(apc,hits,tmpNtuplet.size()+1);
+          for (auto c : tmpNtuplet) cells[c].theTracks.push_back(it);
+          tupleMultiplicity.countDirect(tmpNtuplet.size()+1);
```
Comment (VinInn) [inline RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h]: The timing cost observed is coming from the atomic increment in here, not the condition

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorKernels.cu [hunk 3]
```diff
@@ -185,9 +185,18 @@ void kernel_find_ntuplets(
...
+  __shared__ GPUCACell::TupleMultiplicity::CountersOnly local;
+  if (0==threadIdx.x) local.zero();
+  __syncthreads();
+
   thisCell.find_ntuplets(hh, cells, *foundNtuplets, *apc, *tupleMultiplicity, stack, minHitsPerNtuplet);
   assert(stack.size()==0);
   // printf("in %d found quadruplets: %d\n", cellIndex, apc->get());
+  __syncthreads();
+ 
+  if (0==threadIdx.x) tupleMultiplicity->add(local);
```
Comment (VinInn) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorKernels.cu]: if anything it is slower than directly updating in global

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorKernels.cu [hunk 4]
```diff
@@ -169,22 +169,51 @@ kernel_connect(AtomicPairCounter * apc1, AtomicPairCounter * apc2,  // just to z
...
   if (cellIndex >= (*nCells) ) return;
   auto &thisCell = cells[cellIndex];
-  if (thisCell.theLayerPairId!=0 && thisCell.theLayerPairId!=3 && thisCell.theLayerPairId!=8) return; // inner layer is 0 FIXME
-  GPUCACell::TmpTuple stack;
-  stack.reset();
-  thisCell.find_ntuplets(cells, *foundNtuplets, *apc, stack, minHitsPerNtuplet);
-  assert(stack.size()==0);
-  // printf("in %d found quadruplets: %d\n", cellIndex, apc->get());
+
+  __shared__ GPUCACell::TupleMultiplicity::CountersOnly local;
```
Comment (VinInn) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorKernels.cu]: if anything it is slower than updating in global

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.Xgu7eZWBk9 ...

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.M5srkP3fZE ...

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.1RZGVGalNS ...

**Comment by felicepantaleo** [comment]
:construction: Validation running at vinavx2.cern.ch:/data/user/fpantale/patatrack-validation/run.Yi0tMBiNw4 ...

**Comment by fwyzard** [comment]
@VinInn please rebase this PR on top of CMSSW_10_6_0_pre2 and factor it into independent changes

**Pull request closed without merging**